### PR TITLE
WIP : original LM from TedliumRelease2

### PR DESCRIPTION
**Pull request by vince62s**
This is a modification of the s5_r2 recipe to take into account the LM from Tedlium2 paper.
Gives better results and much better results than in the Paper.
Please review.
Vincent

### CODE REVIEW

**Comment by danpovey** [comment]
@david-ryan-snyder, if you were going to test out the ivector change thing, perhaps you could do it on this setup and kill 2 birds with one stone, making sure this PR runs smoothly?

**Comment by david-ryan-snyder** [comment]
@danpovey, sure, no problem.

**Comment by david-ryan-snyder** [comment]
@danpovey, @vince62s sorry for the delay. I still haven't gotten to this but I will try to do so by the end of the week.

**Comment by danpovey** [comment]
@david-ryan-snyder, don't forget to test this!

**Comment by david-ryan-snyder** [comment]
On it now. Will update ASAP.

File: egs/tedlium/s5_r2/local/ted_train_lm.sh [hunk 1]
```diff
@@ -59,7 +59,8 @@ if [ $stage -le 0 ]; then
   rm ${dir}/data/text/* 2>/dev/null || true
 
   # cantab-TEDLIUM is the larger data source.  gzip it.
```
Comment (david-ryan-snyder) [inline egs/tedlium/s5_r2/local/ted_train_lm.sh]: The comment on line 61 is out of date, isn't it?

File: egs/tedlium/s5_r2/local/ted_train_lm.sh [hunk 2]
```diff
@@ -103,13 +104,7 @@ if [ $stage -le 1 ]; then
...
 
   get_data_prob.py ${dir}/data/real_dev_set.txt ${unpruned_lm_dir} 2>&1 | grep -F '[perplexity'
-
-  # current results, after adding --limit-unk-history=true:
-  # get_data_prob.py: log-prob of data/local/local_lm/data/real_dev_set.txt given model data/local/local_lm/data/wordlist_4.pocolm was -5.13486225358 per word [perplexity = 169.840923284] over 18290.0 words.
-  # older results (after adding min-counts):
-  # get_data_prob.py: log-prob of data/local/local_lm/data/real_dev_set.txt given model data/local/local_lm/data/wordlist_4.pocolm was -5.13902242865 per word [perplexity = 170.514153159] over 18290.0 words.
-  # even older results, before adding min-counts:
-  # get_data_prob.py: log-prob of data/local/local_lm/data/real_dev_set.txt given model data/local/local_lm/data/lm_4 was -5.10576291033 per word [perplexity = 164.969879761] over 18290.0 words.
+  #[perplexity = 157.87] over 18290.0 words
```
Comment (david-ryan-snyder) [inline egs/tedlium/s5_r2/local/ted_train_lm.sh]: The deleted lines appear to have provided a more detailed comment, about the logprob and the model being used. Is it possible to do the same with your updates?

File: egs/tedlium/s5_r2/run.sh [hunk 3]
```diff
@@ -185,7 +185,7 @@ fi
 if [ $stage -le 17 ]; then
   # This will only work if you have GPUs on your system (and note that it requires
   # you to have the queue set up the right way... see kaldi-asr.org/doc/queue.html)
-  local/chain/run_tdnn.sh
+  local/chain/run_tdnn.sh --train-set train --gmm tri3 --nnet3-affix ""
```
Comment (david-ryan-snyder) [inline egs/tedlium/s5_r2/run.sh]: In your RESULTS file, you say that
>  # This is about 0.6% worse than the corresponding results with cleanup.

If that's the case, shouldn't the version with cleanup be the default here (like it was before)?

**Comment by david-ryan-snyder** [comment]
I did something dumb with the PCA vs LDA test I have to rerun something.

In the meantime, I added some comments to the pull request that I hope will be helpful in getting it through.

**Comment by vince62s** [comment]
I will fix the first 2 comments.
the last 2:
no I did not rerun with both min counts perplexities.
the 0.6% worse comment is a bad copy paste from Dan's previous results.
But I did not run the cleaned version.
If you do rerun the whole stuff to update results on a grid, it's better anyway.

**Comment by danpovey** [comment]
@david-ryan-snyder, is this ready to merge?

**Comment by david-ryan-snyder** [comment]
@danpovey,  not quite. @vince62s said he didn't run the cleaned version with his changes. I ran both on the CLSP grid. Hopefully @vince62s can decide what he needs from the following results to complete his RESULTS file.

`%WER 22.5 | 507 17783 | 80.5 14.0 5.5 3.1 22.5 94.7 | 0.092 | exp/tri2/decode_dev/score_15_0.0/ctm.filt.filt.sys
%WER 21.3 | 507 17783 | 81.8 13.1 5.1 3.1 21.3 93.7 | 0.038 | exp/tri2/decode_dev_rescore/score_14_0.0/ctm.filt.filt.sys
%WER 23.6 | 507 17783 | 79.6 14.8 5.6 3.2 23.6 95.1 | 0.024 | exp/tri2/decode_nosp_dev/score_12_0.0/ctm.filt.filt.sys
%WER 22.3 | 507 17783 | 80.7 13.5 5.8 3.0 22.3 93.7 | -0.002 | exp/tri2/decode_nosp_dev_rescore/score_13_0.0/ctm.filt.filt.sys
%WER 23.2 | 1155 27500 | 79.5 15.5 5.0 2.7 23.2 91.1 | 0.070 | exp/tri2/decode_nosp_test/score_12_0.0/ctm.filt.filt.sys
%WER 21.9 | 1155 27500 | 80.7 14.6 4.7 2.6 21.9 90.2 | 0.026 | exp/tri2/decode_nosp_test_rescore/score_12_0.0/ctm.filt.filt.sys
%WER 22.1 | 1155 27500 | 80.7 14.9 4.3 2.8 22.1 90.6 | 0.089 | exp/tri2/decode_test/score_13_0.0/ctm.filt.filt.sys
%WER 20.9 | 1155 27500 | 81.9 14.0 4.1 2.8 20.9 90.5 | 0.046 | exp/tri2/decode_test_rescore/score_13_0.0/ctm.filt.filt.sys
%WER 19.0 | 507 17783 | 83.9 11.4 4.7 2.9 19.0 92.1 | -0.054 | exp/tri3_cleaned/decode_dev/score_13_0.5/ctm.filt.filt.sys
%WER 17.9 | 507 17783 | 85.1 10.5 4.4 3.0 17.9 90.9 | -0.055 | exp/tri3_cleaned/decode_dev_rescore/score_15_0.0/ctm.filt.filt.sys
%WER 22.9 | 507 17783 | 80.0 14.0 5.9 3.0 22.9 94.1 | -0.098 | exp/tri3_cleaned/decode_dev.si/score_14_0.5/ctm.filt.filt.sys
%WER 17.6 | 1155 27500 | 84.8 11.7 3.5 2.4 17.6 87.6 | 0.001 | exp/tri3_cleaned/decode_test/score_15_0.0/ctm.filt.filt.sys
%WER 16.6 | 1155 27500 | 85.8 10.9 3.4 2.4 16.6 86.4 | -0.058 | exp/tri3_cleaned/decode_test_rescore/score_15_0.0/ctm.filt.filt.sys
%WER 22.3 | 1155 27500 | 80.7 15.2 4.1 3.1 22.3 91.0 | -0.092 | exp/tri3_cleaned/decode_test.si/score_13_0.0/ctm.filt.filt.sys
%WER 18.7 | 507 17783 | 83.9 11.4 4.7 2.6 18.7 92.3 | -0.006 | exp/tri3/decode_dev/score_17_0.0/ctm.filt.filt.sys
%WER 17.6 | 507 17783 | 85.0 10.5 4.4 2.6 17.6 90.5 | -0.030 | exp/tri3/decode_dev_rescore/score_16_0.0/ctm.filt.filt.sys
%WER 22.8 | 507 17783 | 80.4 14.1 5.5 3.2 22.8 93.1 | -0.130 | exp/tri3/decode_dev.si/score_12_0.5/ctm.filt.filt.sys
%WER 17.6 | 1155 27500 | 84.7 11.6 3.7 2.4 17.6 87.2 | 0.013 | exp/tri3/decode_test/score_15_0.0/ctm.filt.filt.sys
%WER 16.7 | 1155 27500 | 85.7 10.9 3.4 2.4 16.7 86.4 | -0.044 | exp/tri3/decode_test_rescore/score_14_0.0/ctm.filt.filt.sys
%WER 22.3 | 1155 27500 | 80.6 15.2 4.1 3.0 22.3 91.3 | -0.076 | exp/tri3/decode_test.si/score_13_0.0/ctm.filt.filt.sys
%WER 9.5 | 507 17783 | 91.8 5.8 2.4 1.3 9.5 76.9 | 0.075 | exp/chain_cleaned/tdnn_sp_bi/decode_dev/score_10_0.0/ctm.filt.filt.sys
%WER 8.8 | 507 17783 | 92.4 5.3 2.3 1.3 8.8 76.3 | 0.092 | exp/chain_cleaned/tdnn_sp_bi/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 9.9 | 1155 27500 | 91.4 6.0 2.6 1.3 9.9 74.2 | 0.129 | exp/chain_cleaned/tdnn_sp_bi/decode_test/score_10_0.0/ctm.filt.filt.sys
%WER 9.3 | 1155 27500 | 91.9 5.6 2.5 1.3 9.3 72.3 | 0.100 | exp/chain_cleaned/tdnn_sp_bi/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 10.0 | 507 17783 | 91.3 5.9 2.8 1.3 10.0 78.1 | 0.101 | exp/chain/tdnn_sp_bi/decode_dev/score_9_0.0/ctm.filt.filt.sys
%WER 9.2 | 507 17783 | 92.1 5.3 2.6 1.3 9.2 75.5 | 0.061 | exp/chain/tdnn_sp_bi/decode_dev_rescore/score_9_0.0/ctm.filt.filt.sys
%WER 10.2 | 1155 27500 | 91.0 6.0 3.0 1.2 10.2 75.5 | 0.063 | exp/chain/tdnn_sp_bi/decode_test/score_9_0.0/ctm.filt.filt.sys
%WER 9.8 | 1155 27500 | 91.2 5.4 3.4 1.1 9.8 73.9 | 0.095 | exp/chain/tdnn_sp_bi/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys`

**Comment by david-ryan-snyder** [comment]
@danpovey , the previous results used the normal LDA features for the ivector. Here are results using PCA.

%WER 9.5 | 507 17783 | 91.8 5.8 2.4 1.3 9.5 76.9 | 0.075 | exp/chain_cleaned/tdnn_sp_bi/decode_dev/score_10_0.0/ctm.filt.filt.sys
%WER 8.8 | 507 17783 | 92.4 5.3 2.3 1.3 8.8 76.3 | 0.092 | exp/chain_cleaned/tdnn_sp_bi/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 9.9 | 1155 27500 | 91.4 6.0 2.6 1.3 9.9 74.2 | 0.129 | exp/chain_cleaned/tdnn_sp_bi/decode_test/score_10_0.0/ctm.filt.filt.sys
%WER 9.3 | 1155 27500 | 91.9 5.6 2.5 1.3 9.3 72.3 | 0.100 | exp/chain_cleaned/tdnn_sp_bi/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 10.0 | 507 17783 | 91.3 5.9 2.8 1.3 10.0 78.1 | 0.101 | exp/chain/tdnn_sp_bi/decode_dev/score_9_0.0/ctm.filt.filt.sys
%WER 9.2 | 507 17783 | 92.1 5.3 2.6 1.3 9.2 75.5 | 0.061 | exp/chain/tdnn_sp_bi/decode_dev_rescore/score_9_0.0/ctm.filt.filt.sys
%WER 10.2 | 1155 27500 | 91.0 6.0 3.0 1.2 10.2 75.5 | 0.063 | exp/chain/tdnn_sp_bi/decode_test/score_9_0.0/ctm.filt.filt.sys
%WER 9.8 | 1155 27500 | 91.2 5.4 3.4 1.1 9.8 73.9 | 0.095 | exp/chain/tdnn_sp_bi/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys

**Comment by danpovey** [comment]
not easy to compare as they look different... are the results about the
same?  how do they differ?

On Fri, Nov 25, 2016 at 6:40 PM, david-ryan-snyder <notifications@github.com

**Comment by david-ryan-snyder** [comment]
I added a (vs XX.X) in each line so that you can see the corresponding results with LDA.

Bottom line is, PCA is better in 4 (out of 8) of the results, LDA is better in 3, and they're the same in 1. After averaging all 8 results, PCA is 9.59% and LDA Is 9.63%.

**Comment by danpovey** [comment]
Is this the first time you tested it after your fix?
Maybe we should test on 1 other setup before committing.


On Fri, Nov 25, 2016 at 6:54 PM, david-ryan-snyder <notifications@github.com

**Comment by david-ryan-snyder** [comment]
@danpovey are you referring to the @vince62s or the PCA vs LDA stuff?

**Comment by danpovey** [comment]
No, the PCA vs LDA stuff.

On Fri, Nov 25, 2016 at 7:02 PM, david-ryan-snyder <notifications@github.com

**Comment by david-ryan-snyder** [comment]
There was no fix. The PCA vs LDA results in https://github.com/kaldi-asr/kaldi/issues/1123 use the same setup.

**Comment by david-ryan-snyder** [comment]
> I did something dumb with the PCA vs LDA test I have to rerun something.

@danpovey I was just referring to this PR in particular. The earlier PCA vs LDA results are still valid.

**Comment by david-ryan-snyder** [comment]
@danpovey Since we want to try the PCA vs LDA thing on more recipes, is it OK if we allow @vince62s to finish up the PR without that change? Since we already know the results, it won't need to be rerun later, when we decide to add in PCA for ivectors.

@vince62s I think the main thing is that you need to update your RESULTS file with the cleaned results I posted in an earlier comment. Also, since the cleaned results are better, I imagine you will want to make them the default in the run.sh file (like it was before).

**Comment by danpovey** [comment]
Yes that's fine.

On Fri, Nov 25, 2016 at 7:42 PM, david-ryan-snyder <notifications@github.com

**Comment by vince62s** [comment]
well, what should we do ? just remove all the old LM results or leave them for reference ?

also in @david-ryan-snyder results, there is no "standard" nnet3 results, do I just omit / remove from the file ?

**Comment by david-ryan-snyder** [comment]
I think that's an @danpovey question. 

I can run the other nnet3 results if we need them.

**Comment by danpovey** [comment]
The RESULTS file can have both the new and old results if you want, just
make sure the new ones are displayed more prominently, e.g. first have the
new results, and then say, "below this point are the results before we
changed the LM [describe what you changed]; these old results are more
complete and include neural net results."

File: egs/tedlium/s5_r2/RESULTS [hunk 4]
```diff
@@ -17,50 +21,79 @@
...
+
+# chain+TDNN, large LM
+%WER 9.1 | 507 17783 | 92.3 5.4 2.3 1.3 9.1 76.5 | 0.083 | exp/chain_cleaned/tdnn_sp_bi/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
+%WER 9.3 | 1155 27500 | 91.9 5.6 2.5 1.2 9.3 72.6 | 0.073 | exp/chain_cleaned/tdnn_sp_bi/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
 
+  # chain+TDNN systems ran without cleanup, using the command:
+  # local/chain/run_tdnn.sh --train-set train --gmm tri3 --nnet3-affix ""
+  # for d in exp/chain/tdnn_sp_bi/decode_*; do grep Sum $d/*/*ys | utils/best_wer.sh; done
+  # This is about 0.6% worse than the corresponding results with cleanup.
+AFTER MAX-CHANGE PER COMPONENT
```
Comment (danpovey) [inline egs/tedlium/s5_r2/RESULTS]: Please remove this "AFTER MAX-CHANGE PER COMPONENT" line.. that's history now.

File: egs/tedlium/s5_r2/local/chain/run_tdnn.sh [hunk 5]
```diff
@@ -149,7 +149,7 @@ if [ $stage -le 18 ]; then
     --chain.leaky-hmm-coefficient 0.1 \
     --chain.l2-regularize 0.00005 \
     --chain.apply-deriv-weights false \
-    --chain.lm-opts="--num-extra-lm-states=2000" \
+    --chain.lm-opts="--ngram-order=5 --num-extra-lm-states=2000" \
```
Comment (danpovey) [inline egs/tedlium/s5_r2/local/chain/run_tdnn.sh]: Did you really find that --ngram-order=5 was better?  By how much?
In general I don't like too much tuning that's specific to specific egs directories.  People copy them to other setups, and I prefer to have settings that will work everywhere.

Comment (vince62s) [inline egs/tedlium/s5_r2/local/chain/run_tdnn.sh]: hmm I don't recall exactly. It was very slightly better but not sure how much.
Also I am wondering if this is a change I may have reset after my first commit.
@david-ryan-snyder : in your run it was --ngram-order=5 or just no specified, ie default 4 ?

File: egs/tedlium/s5_r2/RESULTS (hunk 4 above)
Comment (vince62s) [inline egs/tedlium/s5_r2/RESULTS]: right, and the 0.6% worse is no longer 0.6% I'll fix it

File: egs/tedlium/s5_r2/local/chain/run_tdnn.sh (hunk 5 above)
Comment (david-ryan-snyder) [inline egs/tedlium/s5_r2/local/chain/run_tdnn.sh]: I ran it with what you had in the PR, --ngram-order=5.

**Comment by danpovey** [comment]
I tested that option on another setup and did not find it helpful (may have
been a little worse), so decided not to go for that.  Can you please change
it back to 4 and we'll rerun that part?

On Sat, Nov 26, 2016 at 4:12 PM, david-ryan-snyder <notifications@github.com

**Comment by david-ryan-snyder** [comment]
I can rerun it without the --ngram-order=5.

**Comment by vince62s** [comment]
yes go ahead, we'll see how it goes. I just pushed the change.

**Comment by david-ryan-snyder** [comment]
@danpovey, below are the results with ngram-order=4 on the CLSP grid. The average (across both cleaned and regular versions) WER is 9.43% for the ngram-order=4 and 9.63% with ngram-order=5. @vince62s, do you want to update your RESULTS file to reflect this? 

%WER 9.8 | 507 17783 | 91.6 6.0 2.4 1.5 9.8 80.1 | -0.038 | exp/chain/tdnn_sp_bi/decode_dev/score_8_0.0/ctm.filt.filt.sys
%WER 9.1 | 507 17783 | 92.3 5.5 2.3 1.4 9.1 77.5 | 0.011 | exp/chain/tdnn_sp_bi/decode_dev_rescore/score_8_0.0/ctm.filt.filt.sys
%WER 9.9 | 1155 27500 | 91.4 5.7 2.9 1.3 9.9 74.9 | 0.083 | exp/chain/tdnn_sp_bi/decode_test/score_9_0.0/ctm.filt.filt.sys
%WER 9.4 | 1155 27500 | 91.9 5.6 2.5 1.4 9.4 72.7 | 0.018 | exp/chain/tdnn_sp_bi/decode_test_rescore/score_8_0.0/ctm.filt.filt.sys
%WER 9.7 | 507 17783 | 91.7 5.8 2.5 1.4 9.7 78.7 | 0.097 | exp/chain_cleaned/tdnn_sp_bi/decode_dev/score_10_0.0/ctm.filt.filt.sys
%WER 9.0 | 507 17783 | 92.3 5.3 2.4 1.3 9.0 76.7 | 0.067 | exp/chain_cleaned/tdnn_sp_bi/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 9.5 | 1155 27500 | 91.7 5.8 2.5 1.2 9.5 72.5 | 0.079 | exp/chain_cleaned/tdnn_sp_bi/decode_test/score_10_0.0/ctm.filt.filt.sys
%WER 9.0 | 1155 27500 | 92.2 5.3 2.5 1.2 9.0 71.3 | 0.064 | exp/chain_cleaned/tdnn_sp_bi/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys

**Comment by vince62s** [comment]
I will update so but this is annoying how consistently you seem to have better results with n-gram 4 versus me on a single server better with n-gram 5.

**Comment by vince62s** [comment]
ok I think now it should be ok to merge.

File: egs/tedlium/s5_r2/local/chain/run_tdnn.sh [hunk 6]
```diff
@@ -149,7 +149,7 @@ if [ $stage -le 18 ]; then
     --chain.leaky-hmm-coefficient 0.1 \
     --chain.l2-regularize 0.00005 \
     --chain.apply-deriv-weights false \
-    --chain.lm-opts="--num-extra-lm-states=2000" \
+    --chain.lm-opts="--ngram-order=4 --num-extra-lm-states=2000" \
```
Comment (danpovey) [inline egs/tedlium/s5_r2/local/chain/run_tdnn.sh]: can you please remove the --ngram-order=4  since it's the default?  Then should be ready to merge.

**Comment by danpovey** [comment]
Thanks!  Merging.

**Pull request merged**